File: strings/remove_k_digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        if k >= len(num):
            return "0"

        stack = []

        for c in num:
            while stack and k > 0 and stack[-1] > c:
                stack.pop()
                k -= 1
            stack.append(c)

        while k > 0:
            stack.pop()
            k -= 1
        
        result = "".join(stack).lstrip('0')

        return result if result else "0"

File: strings/remove_k_digits.rb
# @param {String} num
# @param {Integer} k
# @return {String}
def remove_kdigits(num, k)
    return "0" if k >= num.size

    stack = []

    num.each_char do |c|
        while !stack.empty? && k > 0 && stack[-1] > c
            stack.pop()
            k -= 1
        end
        stack << c
    end

    while k > 0
        stack.pop
        k -= 1
    end
    
    result = stack.join.sub(/^0+/, '')

    result.empty? ? "0" : result
end